config/initializers/saml_settings.rb

```diff
@@ -1,20 +1,27 @@
 # frozen_string_literal: true
 module SAML
-  CONFIG = Rails.application.config_for(:saml).freeze
-  SETTINGS = OneLogin::RubySaml::Settings.new
+  SAML_CONFIG = Rails.application.config_for(:saml).freeze
```

Comment:
I would keep only this line in the initializer and move everything else into your singleton, you can put them in as local variables with the private fetch method, and since your singleton method for saml_settings memoizes the fetch call, the local variables would only get invoked once.